docs/proposals/kubelet-tls-bootstrap.md

```diff
@@ -0,0 +1,268 @@
+
+Each node is additionally provisioned with the following information:
+
+1. Location of the apiserver
+2. Any CA certificates necessary to trust the apiserver's TLS certificate
+3. Access tokens (if needed) to communicate with the CSR endpoint
+
+These should not change often and are thus simple to include in a static
+provisioning script.
+
+## API Changes
+
+### CertificateSigningRequest Object
+
+We introduce a new API object to represent PKCS#10 certificate signing
+requests. It will be accessible under:
+
+`/apis/certificates/v1beta1/signingrequests/mycsr`
+
+It will have the following structure:
+
+```go
+// Describes a certificate signing request
+type CertificateSigningRequest struct {
+	api.TypeMeta   `json:",inline"`
+	api.ObjectMeta `json:"metadata,omitempty"`
+
+	// The certificate request itself and any additonal information.
+	Spec CertificateSigningRequestSpec `json:"spec,omitempty"`
+
+	// Derived information about the request.
+	Status CertificateSigningRequestStatus `json:"status,omitempty"`
+
+	// The current approval state of the request.
+	Approve CertificateSigningRequestApproval `json:"approve,omitempty"`
+}
+
+// This information is immutable after the request is created.
+type CertificateSigningRequestSpec struct {
+	// base64-encoded PKCS#10 CSR data
+	CertificateRequest string `json:"request"`
+
+	// Any extra information the node wishes to send with the request.
+	ExtraInfo []string `json:"extra,omitempty"`
+}
+
+// This information is derived from the request by Kubernetes and cannot be
+// modified by users. All information is optional since it might not be
+// available in the underlying request. This is intented to aid approval
+// decisions.
+type CertificateSigningRequestStatus struct {
+	// Information about the requesting user (if relevant)
+	// See user.Info interface for details
+	Username string   `json:"username,omitempty"`
+	UID      string   `json:"uid,omitempty"`
+	Groups   []string `json:"groups,omitempty"`
+
+	// Fingerprint of the public key in request
+	Fingerprint string `json:"fingerprint,omitempty"`
+
+	// Subject fields from the request
+	Subject pkix.Name `json:"subject,omitempty"`
+
+	// DNS SANs from the request
+	Hostnames []string `json:"dns,omitempty"`
+
+	// IP SANs from the request
+	IPAddresses []string `json:"ip,omitempty"`
+}
+
+type CertificateSigningRequestApproval struct {
+	// CSR approval state, one of Submitted, Approved, or Denied
+	State CertificateRequestState `json:"state"`
+
+	// brief reason for the request state
+	Reason string `json:"reason,omitempty"`
+	// human readable message with details about the request state
+	Message string `json:"message,omitempty"`
+
+	// If request was approved, the controller will place the issued certificate here.
+	Certificate []byte `json:"certificate,omitempty"`
+}
+
+type CertificateRequestState string
+
+// These are the possible states for a certificate request.
+const (
+	RequestSubmitted CertificateRequestState = "Submitted"
```

Comment:
Submitted is intended to express that the CSR has passed validation logic (which currently just checks it for a valid signature).